crypto/bn/asm/rsaz-avx512.pl

```diff
@@ -0,0 +1,706 @@
+    vpmadd52huq `160*$idx+64*1+32`($a), $Bi, $R1h
+    vpmadd52huq `160*$idx+64*2`($a), $Bi, $R2
+
+    vpmadd52huq `160*$idx+64*0`($m), $Yi, $R0
+    vpmadd52huq `160*$idx+64*0+32`($m), $Yi, $R0h
+    vpmadd52huq `160*$idx+64*1`($m), $Yi, $R1
+    vpmadd52huq `160*$idx+64*1+32`($m), $Yi, $R1h
+    vpmadd52huq `160*$idx+64*2`($m), $Yi, $R2
+___
+}
+
+# Uses %r8-14
+sub amm52x20_x1_norm {
+my ($acc,$R0,$R0h,$R1,$R1h,$R2) = @_;
+$code.=<<___;
+    vpbroadcastq    $acc, $TMP
+    vpblendd \$3, $TMP, $R0, $R0
+    vpsrlq    \$52, $R0,   $T1
+    vpsrlq    \$52, $R0h,  $T1h
+    vpsrlq    \$52, $R1,   $T2
+    vpsrlq    \$52, $R1h,  $T2h
+    vpsrlq    \$52, $R2,   $T3
+    valignq \$3, $T2h,  $T3,  $T3
+    valignq \$3, $T2,   $T2h, $T2h
+    valignq \$3, $T1h,  $T2,  $T2
+    valignq \$3, $T1,   $T1h, $T1h
+    valignq \$3, $zero, $T1,  $T1
+    vpandq    $mask52x4, $R0,  $R0
+    vpandq    $mask52x4, $R0h, $R0h
+    vpandq    $mask52x4, $R1,  $R1
+    vpandq    $mask52x4, $R1h, $R1h
+    vpandq    $mask52x4, $R2,  $R2
+    vpaddq  $T1,  $R0,  $R0
+    vpaddq  $T2,  $R1,  $R1
+    vpaddq  $T3,  $R2,  $R2
+    vpaddq  $T1h, $R0h, $R0h
+    vpaddq  $T2h, $R1h, $R1h
+    vpcmpuq   \$1, $R0,  $mask52x4, %k1 # OP=lt
+    vpcmpuq   \$1, $R1,  $mask52x4, %k2
+    vpcmpuq   \$1, $R2,  $mask52x4, %k3
+    vpcmpuq   \$0, $R0,  $mask52x4, %k4 # OP=eq
+    vpcmpuq   \$0, $R1,  $mask52x4, %k5
+    vpcmpuq   \$0, $R2,  $mask52x4, %k6
+    kmovb   %k1, %r14d
+    kmovb   %k2, %r13d
+    kmovb   %k3, %r12d
+    kmovb   %k4, %r11d
+    kmovb   %k5, %r10d
+    kmovb   %k6, %r9d
+    leal    (%r11,%r14,2), %r11d        # k4 += 2*k1
+    leal    (%r10,%r13,2), %r10d        # k5 += 2*k2
+    leal    (%r9,%r12,2), %r9d          # k6 += 2*k3
+    xorl    %r11d, %r14d                # k1 ^= k4
+    xorl    %r10d, %r13d                # k2 ^= k5
+    xorl    %r9d, %r12d                 # k3 ^= k6
+    vpcmpuq   \$1, $R0h, $mask52x4, %k1 # OP=lt
+    vpcmpuq   \$1, $R1h, $mask52x4, %k2
+    vpcmpuq   \$0, $R0h, $mask52x4, %k3 # OP=eq
+    vpcmpuq   \$0, $R1h, $mask52x4, %k4
+    kmovb   %k1, %r11d                  # k1h
+    kmovb   %k2, %r10d                  # k2h
+    kmovb   %k3, %r9d                   # k4h
+    kmovb   %k4, %r8d                   # k5h
+    leal    (%r9,%r11,2), %r9d          # k4h += 2*k1h
+    leal    (%r8,%r10,2), %r8d          # k5h += 2*k2h;
+    xorl    %r9d, %r11d                 # k1h ^= k4h
+    xorl    %r8d, %r10d                 # k2h ^= k5h
+    kmovb   %r14d, %k1
+    kmovb   %r13d, %k2
+    kmovb   %r12d, %k3
+    kmovb   %r11d, %k4
+    kmovb   %r10d, %k5
+    vpsubq  $mask52x4, $R0,  ${R0}{%k1}
+    vpsubq  $mask52x4, $R1,  ${R1}{%k2}
+    vpsubq  $mask52x4, $R2,  ${R2}{%k3}
+    vpsubq  $mask52x4, $R0h, ${R0h}{%k4}
+    vpsubq  $mask52x4, $R1h, ${R1h}{%k5}
+    vpandq   $mask52x4, $R0, $R0
+    vpandq   $mask52x4, $R1, $R1
+    vpandq   $mask52x4, $R2, $R2
+    vpandq   $mask52x4, $R0h, $R0h
+    vpandq   $mask52x4, $R1h, $R1h
```

Comment:
There's something I don't understand here, 

After line 405, there may still be carries across QWs to handle, which is done between 406 and 450.
Before entering line 406, there may be a carry from one 256-bit register which needs to ripple into another. For example, imagine the most-significant QW of $R0 is 53-bits (which is possible as a result of the previous normalization step). Then this top bit (bit-52) must be added to the least significant QW of $R0_0h.

From the code between 406 and 450, it looks to be handling carry propagation within each 4 QW register, but I can't see how the carry propagation between each register is happening.

At line 405, with random data it is quite unlikely to have any carries to propagate at this stage, since it means that adding a number of only a few bits to a 64-bit number results in a number > 2^64, which has a probability of something like 2^-62 of happening with random data. But if it can happen, then it must be dealt with properly, otherwise it is a ticking time bomb.

Apologies if there is some subtlety in the assembly which I'm not seeing.

Reply:
Also, something about the sequence of operations on the masks which I don't understand.

Taking $R0 as an example, there is a sequence as follows in the code..

k4 += 2*k1
k1 ^= k4

it looks as though it should be..

k1 = k4 + 2*k1
k1 ^= k4

to ensure the correct QWs are adjusted

Reply:
There're alignr's that puts corresponding carries to appropriate QWs positions. So carries propagated between registers here. I also added more comments through the function.
As to mask ops comment - you seem to be right, I've updated this piece of code.

Reply:
With random data I think issues like this will rarely be exposed.

Since it very unlikely to hit all the corner cases of this routine with random data, if not done already it would be good to test the updated routine in isolation with carefully chosen inputs. For example, using input {x[0] = 2^52, x[1..18] = 2^52 - 1, x[19] = 0}. Then it can verify that a carry gets propagated from x[0] all the way to x[19], crossing $ymm register boundaries and QW boundaries within each $ymm register

Reply:
I have updated this piece of code again, there was a bug. Now it should process carries as expected. An isolated testing for _norm procedure with predefined inputs revealed no issues. Thanks for your comments.

Reply:
Thanks, the updates make sense.

There is 1 thing to check - see the review comments for the latest commit